File: e2e/java/src/test/java/dev/kreuzberg/e2e/MarkdownTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import dev.kreuzberg.kreuzcrawl.Kreuzcrawl;

/** E2e tests for category: markdown. */
class MarkdownTest {
    @Test
    void testMarkdownBasicConversion() throws Exception {
        // HTML is always converted to markdown alongside raw HTML
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        assertEquals("Test", result.metadata().title().orElse(""));
        assertFalse(result.html().isEmpty(), "expected non-empty value");
        assertFalse(result.markdown().orElseThrow().content().isEmpty(), "expected non-empty value");
        assertTrue(result.markdown().orElseThrow().content().contains("Hello World"), "expected to contain: " + "Hello World");
    }

    @Test
    void testMarkdownCrawlAllPages() throws Exception {
        // All crawled pages have markdown field populated
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        // skipped: field 'crawl.pages_crawled' not available on result type
    }

    @Test
    void testMarkdownFitContent() throws Exception {
        // Fit markdown removes navigation and boilerplate content
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        assertFalse(result.markdown().orElseThrow().content().isEmpty(), "expected non-empty value");
    }

    @Test
    void testMarkdownHeadingsAndParagraphs() throws Exception {
        // Markdown conversion preserves heading hierarchy and paragraph text
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertFalse(result.markdown().orElseThrow().content().isEmpty(), "expected non-empty value");
        assertTrue(result.markdown().orElseThrow().content().contains("Main Title"), "expected to contain: " + "Main Title");
    }

    @Test
    void testMarkdownLinksConverted() throws Exception {
        // HTML links are converted to markdown link syntax
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        assertFalse(result.html().isEmpty(), "expected non-empty value");
        assertFalse(result.markdown().orElseThrow().content().isEmpty(), "expected non-empty value");
        assertTrue(result.markdown().orElseThrow().content().contains("Example"), "expected to contain: " + "Example");
    }

    @Test
    void testMarkdownWithCitations() throws Exception {
        // Markdown includes citation conversion with numbered references
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        assertFalse(result.markdown().orElseThrow().content().isEmpty(), "expected non-empty value");
    }

}

File: e2e/go/markdown_test.go
// E2e tests for category: markdown
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_MarkdownBasicConversion(t *testing.T) {
	// HTML is always converted to markdown alongside raw HTML
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var metadata_title string
	if result.Metadata.Title != nil {
		metadata_title = *result.Metadata.Title
	}
	var markdown string
	if result.Markdown != nil {
		markdown = *result.Markdown
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %q", result.StatusCode)
	}
	if metadata_title != `Test` {
		t.Errorf("equals mismatch: got %q", metadata_title)
	}
	if len(result.Html) == 0 {
		t.Errorf("expected non-empty value")
	}
	if len(markdown) == 0 {
		t.Errorf("expected non-empty value")
	}
	if !strings.Contains(markdown, `Hello World`) {
		t.Errorf("expected to contain %s, got %q", `Hello World`, markdown)
	}
}

func Test_MarkdownCrawlAllPages(t *testing.T) {
	// All crawled pages have markdown field populated
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'crawl.pages_crawled' not available on result type
}

func Test_MarkdownFitContent(t *testing.T) {
	// Fit markdown removes navigation and boilerplate content
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var markdown string
	if result.Markdown != nil {
		markdown = *result.Markdown
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %q", result.StatusCode)
	}
	if len(markdown) == 0 {
		t.Errorf("expected non-empty value")
	}
}

func Test_MarkdownHeadingsAndParagraphs(t *testing.T) {
	// Markdown conversion preserves heading hierarchy and paragraph text
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var markdown string
	if result.Markdown != nil {
		markdown = *result.Markdown
	}
	if len(markdown) == 0 {
		t.Errorf("expected non-empty value")
	}
	if !strings.Contains(markdown, `Main Title`) {
		t.Errorf("expected to contain %s, got %q", `Main Title`, markdown)
	}
}

func Test_MarkdownLinksConverted(t *testing.T) {
	// HTML links are converted to markdown link syntax
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var markdown string
	if result.Markdown != nil {
		markdown = *result.Markdown
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %q", result.StatusCode)
	}
	if len(result.Html) == 0 {
		t.Errorf("expected non-empty value")
	}
	if len(markdown) == 0 {
		t.Errorf("expected non-empty value")
	}
	if !strings.Contains(markdown, `Example`) {
		t.Errorf("expected to contain %s, got %q", `Example`, markdown)
	}
}

func Test_MarkdownWithCitations(t *testing.T) {
	// Markdown includes citation conversion with numbered references
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var markdown string
	if result.Markdown != nil {
		markdown = *result.Markdown
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %q", result.StatusCode)
	}
	if len(markdown) == 0 {
		t.Errorf("expected non-empty value")
	}
}
